File: leetcode3/CountNodesEqualtoAverageofSubtree/a.ts
// Copyright (C) 2023 by iamslash

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}

// 53ms 100.00% 47.9MB 42.86%
// dfs
// O(N) O(lgN)
function averageOfSubtree(root: TreeNode | null): number {
    let best = 0;
    function dfs(u: TreeNode | null): [number, number] {
        // base
        if (u == null) {
            return [0, 0];
        }
        // recursion
        const [lsum, lcnt] = dfs(u.left)
        const [rsum, rcnt] = dfs(u.right)
        const sum = lsum + rsum + u.val
        const cnt = lcnt + rcnt + 1
        if (Math.floor(sum / cnt) === u.val) {
            best++;
        }
        return [sum, cnt]
    }
    dfs(root)
    return best
};

File: leetcode3/CountNodesEqualtoAverageofSubtree/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2023 by iamslash

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

        # 44ms 77.02% 15.1MB 67.65%
# dfs
# O(N) O(lgN)
class Solution:
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:
        best = 0

        def dfs(u):
            nonlocal best
            # base
            if u == None:
                return 0, 0
            
            # recursion
            lsum, lcnt = dfs(u.left)
            rsum, rcnt = dfs(u.right)
            sum = lsum + rsum + u.val
            cnt = lcnt + rcnt + 1
            if sum // cnt == u.val:
                best += 1

            return sum, cnt    
                
        dfs(root)
        return best

